src/app/archive/lib/processor.ml

```diff
@@ -1417,20 +1421,119 @@ module Zkapp_events = struct
 
   let table_name = "zkapp_events"
 
+  let sep_by_comma ?(parenthesis = false) xs =
+    List.map xs ~f:(if parenthesis then sprintf "('%s')" else sprintf "'%s'")
+    |> String.concat ~sep:", "
+
+  (* Account_update.Body.Events'.t is defined as `field array list`,
+     which is ismorphic to a list of list of fields.
+
+     We are batching the insertion of field and field_array to optimize
+     the speed of archiving max-cost zkapps.
+
+     1. we flatten the list of list of fields to get all the field elements
+     2. insert all the field elements in one query
+     3. construct a map "M" from `field_id` to `field` by querying against the zkapp_field table
+     4. use "M" and the list of list of fields to compute the list of list of field_ids
+     5. insert all list of `list of field_ids` in one query
+     6. construct a map "M'" from `field_array_id` to `field_id array` by querying against
+        the zkapp_field_array table
+     7. use "M'" and the list of list of field_ids to compute the list of field_array_ids
+     8. insert the list of field_arrays
+  *)
   let add_if_doesn't_exist (module Conn : CONNECTION)
       (events : Account_update.Body.Events'.t) =
     let open Deferred.Result.Let_syntax in
-    let%bind (element_ids : int array) =
-      Mina_caqti.deferred_result_list_map events
-        ~f:(Zkapp_field_array.add_if_doesn't_exist (module Conn))
-      >>| Array.of_list
+    let%bind field_array_id_list =
+      if not @@ List.is_empty events then
+        let field_list_list =
+          List.map events ~f:(fun field_array ->
+              Array.map field_array ~f:Pickles.Backend.Tick.Field.to_string
+              |> Array.to_list )
+        in
+        let fields = field_list_list |> List.concat in
+        let%bind field_id_list_list =
+          if not @@ List.is_empty fields then
+            let field_insert =
+              sprintf
```

Comment:
Do you think we can generalize the pattern and extract it as a separate function to reduce code duplication?

I'm taking about `insert on conflict do nothing` paired with select of `some_field, id` with conversion of it to map preserving the initial order.

Reply:
Yeah. Let me add a function in mina_caqti for that.

Reply:
Nevermind, I got most part of the refactor done. I left out the conversion from value to string.